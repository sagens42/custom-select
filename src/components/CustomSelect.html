<button aria-haspopup="listbox">
</button>
<ul
  class="custom-options hidden"
  role="listbox"
  tabindex="-1"
  >
</ul>
