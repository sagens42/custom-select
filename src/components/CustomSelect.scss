.custom-select-wrapper {
  display: inline-block;

  ul.hidden {
    display: none;
  }

  button {
    border-radius: 5px;
    font-size: 16px;
    text-align: left;
    padding: 5px 10px;
    width: 150px;
    position: relative;
  }

  button:after {
    width: 0;
    height: 0;
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-top: 8px solid #aaa;
    content: " ";
    position: absolute;
    right: 5px;
    top: 10px;
  }

  ul {
    border-top: 0;
    max-height: 10em;
    overflow-y: auto;
    position: absolute;
    margin: 0;
    width: 148px;
    padding: 0;
    background: white;
    border: 1px solid #aaa;
  }

  ul li {
    display: block;
    padding: 5px 10px;
    position: relative;

    &.active {
      background: #bde4ff;
    }
  }
}
