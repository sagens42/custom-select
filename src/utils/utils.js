function createElement(html) {
  var div = document.createElement('div');
  div.innerHTML = html.trim();
  return div.firstChild;
}

function createEvent(type) {
  var event = document.createEvent('Event');
  event.initEvent(type);
  return event;
}

export { createElement, createEvent };
