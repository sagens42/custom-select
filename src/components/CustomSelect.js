import './CustomSelect.scss';
import markup from './CustomSelect.html';
import markupForMultiple from './CustomSelectMultiple.html';
import { createElement, createEvent } from '../utils/utils';

export default class CustomSelect {
  constructor(select) {
    this.select = select;
    this.multiple = select.multiple || false;
    this.keydownIndex = this.multiple ? null : 0;
    this.customSelectWrapper = document.createElement('div');
    this.customSelectWrapper.classList.add('custom-select-wrapper');
    if (this.multiple) {
      this.customSelectWrapper.innerHTML = markupForMultiple;
    } else {
      this.customSelectWrapper.innerHTML = markup;
      this.customSelectButton = this.customSelectWrapper.querySelector('button');
    }
    this.customSelectList = this.customSelectWrapper.querySelector('ul');
    // check for assosiated label
    var labeledBy = this.select.getAttribute('aria-labelledby');
    if (!labeledBy) {
      var label = document.querySelector(`label[for="${select.id}"]`);
      if (label) {
        labeledBy = label.id;
        this.customSelectButton.setAttribute('aria-labelledby', label.id);
        this.customSelectList.setAttribute('aria-labelledby', label.id);
      }
    } else {
      this.customSelectButton.setAttribute('aria-labelledby', labeledBy);
      this.customSelectList.setAttribute('aria-labelledby', labeledBy);
    }

    this.setupOptions();
    if (!this.multiple) {
      this.customSelectButton.textContent = this.select.value;
    }
    this.bindEvents();
    this.customSelectWrapper.instance = this;

    return this.customSelectWrapper;
  }

  getSelectedValues() {
    if (this.multiple) {
      const result = [];
      const options = this.select.options;
      for (let i = 0; i < options.length; i++) {
        const option = options[i];

        if (option.selected) {
          result.push(option.value || option.text);
        }
      }
      return result;
    } else {
      return this.select.value;
    }
  }

  setupOptions() {
    // setting original options
    const customSelectOptions = this.customSelectWrapper.querySelector('.custom-options');
    for (let i = 0; i < this.select.options.length; i++) {
      const customSelectOption = document.createElement('li');
      customSelectOption.classList.add('custom-option');
      customSelectOption.setAttribute('role', 'option');
      customSelectOption.textContent = this.select.options[i].textContent;

      if (this.multiple) {
        if (this.getSelectedValues().includes(this.select.options[i].textContent)) {
          customSelectOption.classList.add('active');
          customSelectOption.setAttribute('aria-selected', 'true');
        }
      } else {
        if (this.select.options[i].textContent.toLowerCase() === this.select.value.toLowerCase()) {
          customSelectOption.classList.add('active');
          customSelectOption.setAttribute('aria-selected', 'true');
        }
      }

      customSelectOption.addEventListener('click', this.onClickOnOption.bind(this));

      customSelectOptions.appendChild(customSelectOption);
    }
  }

  toggleDropdown() {
    const hidden = this.customSelectList.classList.contains('hidden');
    this.customSelectButton.parentNode.querySelector('ul').classList.toggle('hidden');
    if (hidden) {
      this.customSelectButton.setAttribute('aria-expanded', 'true');
    } else {
      this.customSelectButton.removeAttribute('aria-expanded');
    }
  }

  onClickOnOption(e) {
    const prevActive = this.customSelectWrapper.querySelector('.custom-option.active');
    if (this.multiple) {
      if (!e.metaKey && !e.ctrlKey && prevActive) {
        this.deselectItem(prevActive);
      }
      if (e.target.classList.contains('active')) {
        this.deselectItem(e.target);
      } else {
        this.selectItem(e.target);
      }
    } else {
      if (!e.target.classList.contains('active')) {
        if (prevActive) {
          this.deselectItem(prevActive);
        }
        this.selectItem(e.target);
        this.toggleDropdown();
      }
    }
    this.triggerEventOnElement(this.select, 'change');
  }

  selectItem(item) {
    item.classList.add('active');
    item.setAttribute('aria-selected', 'true');
    if (this.multiple) {
      const options = this.select.options;
      for (let i = 0; i < options.length; i++) {
        const option = options[i];
        if (option.textContent.toLowerCase() === item.textContent.toLowerCase()) {
          option.selected = true;
        }
      }
    } else {
      this.customSelectButton.textContent = item.textContent;
      this.select.value = item.textContent;
    }
  }

  deselectItem(item) {
    item.classList.remove('active');
    item.removeAttribute('aria-selected');
  }

  deselectAllItems() {
    const elements = this.customSelectWrapper.querySelectorAll('li');
    for (let i = 0; i < elements.length; i++) {
      this.deselectItem(elements[i]);
    }
  }

  findIndexOfElement(element) {
    const elements = this.customSelectWrapper.querySelectorAll('li');
    for (let i = 0; i < elements.length; i++) {
      if (element.textContent.toLowerCase() === elements[i].textContent.toLowerCase()) {
        return i;
      }
    }
  }

  detectKeyDown(e) {
    if (!this.customSelectWrapper.querySelector('ul').classList.contains('hidden')) {
      const selectedItem = this.keydownIndex != null ? this.customSelectWrapper.querySelectorAll('li')[this.keydownIndex] : null;
      if ((e.which || e.keyCode) === 40) { // down
        if (selectedItem) {
          const nextItem = selectedItem.nextElementSibling;
          if (nextItem) {
            if (this.multiple) {
              if (!e.metaKey && !e.ctrlKey) {
                this.deselectAllItems();
              }
            } else {
              this.deselectItem(selectedItem);
            }
            this.selectItem(nextItem);
            this.keydownIndex = this.findIndexOfElement(nextItem);
          }
        } else {
          const firstItem = this.customSelectWrapper.querySelector('li');
          if (firstItem) {
            this.selectItem(firstItem);
            this.keydownIndex = 0;
          }
        }
    } else if ((e.which || e.keyCode) === 38) { // up
        if (selectedItem) {
          const previousItem = selectedItem.previousElementSibling;
          if (previousItem) {
            if (this.multiple) {
              if (!e.metaKey && !e.ctrlKey) {
                this.deselectAllItems();
              }
            } else {
              this.deselectItem(selectedItem);
            }
            this.selectItem(previousItem);
            this.keydownIndex = this.findIndexOfElement(previousItem);
          }
        }
      } else if ((e.which || e.keyCode) === 9) { // tab
        if (!this.multiple) {
          this.toggleDropdown();
        }
      }
    }
  }

  bindEvents() {
    // setting events
    if (!this.multiple) {
      this.customSelectButton.addEventListener('click', this.toggleDropdown.bind(this));
      this.customSelectButton.addEventListener('keydown', this.detectKeyDown.bind(this));
    } else {
      this.customSelectList.addEventListener('keydown', this.detectKeyDown.bind(this));
    }

    // trigger main events on original select just in case if developer binds custom events to origincal select
    // this.customSelectWrapper.addEventListener('click', this.triggerEventOnOriginalSelect.bind(this));
    // this.customSelectWrapper.addEventListener('keydown', this.triggerEventOnOriginalSelect.bind(this));
    // this.customSelectWrapper.addEventListener('keyup', this.triggerEventOnOriginalSelect.bind(this));
    // this.customSelectWrapper.addEventListener('focus', this.triggerEventOnOriginalSelect.bind(this));
    // this.customSelectWrapper.addEventListener('blur', this.triggerEventOnOriginalSelect.bind(this));
  }

  triggerEventOnOriginalSelect(e) {
    this.select.dispatchEvent(e);
  }

  triggerEventOnElement(element, eventType) {
    const event = createEvent(eventType);
    this.select.dispatchEvent(event);
  }
}
