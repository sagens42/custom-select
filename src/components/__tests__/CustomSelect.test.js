import CustomSelect from '../CustomSelect';
import { createElement, createEvent } from '../../utils/utils';

let selectTemplate = `
  <select>
    <option>One</option>
    <option>Two</option>
    <option>Three</option>
  </select>
`;

describe('Single CustomSelect', () => {
  let singleSelect;
  beforeEach(() => {
    let select = createElement(selectTemplate);
    singleSelect = new CustomSelect(select);
  });

  it('should init component', () => {
    expect(singleSelect.instance.select).not.toBeNull();
    expect(singleSelect.instance.multiple).toBeFalsy();
    expect(singleSelect.instance.keydownIndex).toBe(0);
    expect(singleSelect).toMatchSnapshot();
  });

  it('should be able to select items', () => {
    singleSelect.instance.toggleDropdown();
    expect(singleSelect.instance.customSelectList.classList).not.toContain('hidden');

    singleSelect.instance.customSelectWrapper.querySelector('li').click();
    expect(singleSelect.instance.customSelectWrapper.querySelector('li').classList).toContain('active');

    singleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].click();
    expect(singleSelect.instance.customSelectWrapper.querySelector('li').classList).not.toContain('active');
    expect(singleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].classList).toContain('active');
  });

  it('should be able navigate with keyboard', () => {
    // pressing down
    const downEvent = createEvent('keydown');
    downEvent.keyCode = 40;
    singleSelect.instance.toggleDropdown();
    expect(singleSelect.instance.customSelectList.classList).not.toContain('hidden');
    expect(singleSelect.instance.customSelectWrapper.querySelector('li').classList).toContain('active');

    singleSelect.instance.customSelectButton.dispatchEvent(downEvent);
    expect(singleSelect.instance.customSelectWrapper.querySelector('li').classList).not.toContain('active');
    expect(singleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].classList).toContain('active');

    // pressing up
    const upEvent = createEvent('keydown');
    upEvent.keyCode = 38;
    singleSelect.instance.customSelectButton.dispatchEvent(upEvent);
    expect(singleSelect.instance.customSelectWrapper.querySelector('li').classList).toContain('active');
    expect(singleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].classList).not.toContain('active');

    // pressing tab
    const tabEvent = createEvent('keydown');
    tabEvent.keyCode = 9;
    singleSelect.instance.customSelectButton.dispatchEvent(tabEvent);
    expect(singleSelect.instance.customSelectList.classList).toContain('hidden');
  });

  it('should trigger original onchange event on change', () => {
    window.callMe = jest.fn(() => {});
    let selectTemplate = `
      <select onchange="callMe()">
        <option>One</option>
        <option>Two</option>
        <option>Three</option>
      </select>
    `;
    const select = createElement(selectTemplate);
    const customSelect = new CustomSelect(select);
    customSelect.instance.customSelectWrapper.querySelector('li').click();
    expect(window.callMe).toBeCalled();
    expect(select.value).toBe('One');
  });
});

describe('Multiple CustomSelect', () => {
  let multipleSelect;
  beforeEach(() => {
    let select = createElement(selectTemplate);
    select.multiple = true;
    multipleSelect = new CustomSelect(select);
  });

  it('should init component', () => {
    expect(multipleSelect.instance.select).not.toBeNull();
    expect(multipleSelect.instance.multiple).toBeTruthy();
    expect(multipleSelect.instance.keydownIndex).toBeNull();
    expect(multipleSelect).toMatchSnapshot();
  });

  it('should be able to select items', () => {
    multipleSelect.instance.customSelectWrapper.querySelector('li').click();
    expect(multipleSelect.instance.customSelectWrapper.querySelector('li').classList).toContain('active');

    multipleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].click();
    expect(multipleSelect.instance.customSelectWrapper.querySelector('li').classList).not.toContain('active');
    expect(multipleSelect.instance.customSelectWrapper.querySelectorAll('li')[1].classList).toContain('active');
  });

  it('should be able navigate with keyboard', () => {
    const options = multipleSelect.instance.customSelectWrapper.querySelectorAll('li');
    // pressing down
    const downEvent = createEvent('keydown');
    downEvent.keyCode = 40;

    multipleSelect.instance.customSelectList.dispatchEvent(downEvent);
    // double down event to be able to test ctrl + up below
    multipleSelect.instance.customSelectList.dispatchEvent(downEvent);
    expect(options[0].classList).not.toContain('active');
    expect(options[1].classList).toContain('active');
    expect(options[2].classList).not.toContain('active');

    // pressing down with ctrl
    downEvent.ctrlKey = true;
    multipleSelect.instance.customSelectList.dispatchEvent(downEvent);
    expect(options[0].classList).not.toContain('active');
    expect(options[1].classList).toContain('active');
    expect(options[2].classList).toContain('active');

    // pressing up
    const upEvent = createEvent('keydown');
    upEvent.keyCode = 38;
    multipleSelect.instance.customSelectList.dispatchEvent(upEvent);
    expect(options[0].classList).not.toContain('active');
    expect(options[1].classList).toContain('active');
    expect(options[2].classList).not.toContain('active');

    // pressing up with ctrl
    upEvent.ctrlKey = true;
    multipleSelect.instance.customSelectList.dispatchEvent(upEvent);
    expect(options[0].classList).toContain('active');
    expect(options[1].classList).toContain('active');
    expect(options[2].classList).not.toContain('active');
  });

  it('should trigger original onchange event on change', () => {
    window.callMe = jest.fn(() => {});
    let selectTemplate = `
      <select multiple onchange="callMe()">
        <option>One</option>
        <option>Two</option>
        <option>Three</option>
      </select>
    `;
    const select = createElement(selectTemplate);
    const customSelect = new CustomSelect(select);
    customSelect.instance.customSelectWrapper.querySelector('li').click();
    expect(window.callMe).toBeCalled();
    expect(select.value).toBe('One');
  });
});
