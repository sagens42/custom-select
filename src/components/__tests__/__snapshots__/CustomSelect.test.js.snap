// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple CustomSelect should init component 1`] = `
<div
  class="custom-select-wrapper"
>
  <ul
    class="custom-options"
    role="listbox"
    tabindex="0"
  >
    

    <li
      aria-selected="true"
      class="custom-option active"
      role="option"
    >
      One
    </li>
    <li
      class="custom-option"
      role="option"
    >
      Two
    </li>
    <li
      class="custom-option"
      role="option"
    >
      Three
    </li>
  </ul>
  

</div>
`;

exports[`Single CustomSelect should init component 1`] = `
<div
  class="custom-select-wrapper"
>
  <button
    aria-haspopup="listbox"
  >
    One
  </button>
  

  <ul
    class="custom-options hidden"
    role="listbox"
    tabindex="-1"
  >
    

    <li
      aria-selected="true"
      class="custom-option active"
      role="option"
    >
      One
    </li>
    <li
      class="custom-option"
      role="option"
    >
      Two
    </li>
    <li
      class="custom-option"
      role="option"
    >
      Three
    </li>
  </ul>
  

</div>
`;
